topology/probes/k8s/container.go

```diff
@@ -37,82 +43,101 @@ const (
-		containers := p.handleContainers(node)
+// OnAdd is called when a new Kubernetes resource has been created
+func (c *containerCache) OnAdd(obj interface{}) {
+	if pod, ok := obj.(*v1.Pod); ok {
+		c.graph.Lock()
+		defer c.graph.Unlock()
+
+		wasUpdated := make(map[string]bool)
+
+		for _, container := range pod.Spec.Containers {
+			uid := c.genID(pod, container.Name)
+			m := c.newMetadata(pod, &container)
+			node := c.graph.GetNode(uid)
+			if node == nil {
+				node = c.graph.NewNode(uid, m)
+				c.NotifyEvent(graph.NodeAdded, node)
+				logging.GetLogger().Debugf("Added %s", c.dump(pod, container.Name))
+			} else {
+				c.graph.SetMetadata(node, m)
+				c.NotifyEvent(graph.NodeUpdated, node)
+				logging.GetLogger().Debugf("Updated %s", c.dump(pod, container.Name))
+			}
+			wasUpdated[container.Name] = true
+		}
 
-		for _, container := range previousContainers {
-			if _, found := containers[container.ID]; !found {
-				p.graph.DelNode(container)
+		nodes, _ := c.containerIndexer.Get(pod.Namespace, pod.Name)
+		for _, node := range nodes {
+			name, _ := node.GetFieldString("Name")
+			if !wasUpdated[name] {
+				c.graph.DelNode(node)
+				c.NotifyEvent(graph.NodeDeleted, node)
+				logging.GetLogger().Debugf("Deleted %s", c.dump(pod, name))
 			}
 		}
 	}
 }
 
-func (p *containerProbe) OnNodeDeleted(node *graph.Node) {
-	if nodeType, _ := node.GetFieldString("Type"); nodeType == "pod" {
-		containers := p.graph.LookupChildren(node, graph.Metadata{"Type": "container"}, topology.OwnershipMetadata())
-		for _, container := range containers {
-			p.graph.DelNode(container)
+// OnUpdate is called when a Kubernetes resource has been updated
+func (c *containerCache) OnUpdate(oldObj, newObj interface{}) {
+	c.OnAdd(newObj)
+}
+
+// OnDelete is called when a Kubernetes resource has been deleted
+func (c *containerCache) OnDelete(obj interface{}) {
+	if pod, ok := obj.(*v1.Pod); ok {
+		c.graph.Lock()
+		defer c.graph.Unlock()
+
+		containerNodes, _ := c.containerIndexer.Get(pod.Namespace, pod.Name)
+		for _, containerNode := range containerNodes {
+			name, _ := containerNode.GetFieldString("Name")
+			c.graph.DelNode(containerNode)
+			c.NotifyEvent(graph.NodeDeleted, containerNode)
+			logging.GetLogger().Debugf("Deleted %s", c.dump(pod, name))
 		}
 	}
 }
 
-func (p *containerProbe) Start() {
-	p.graph.AddEventListener(p)
+func newContainerCache(restClient rest.Interface, objType runtime.Object, resources string, g *graph.Graph) *containerCache {
+	c := &containerCache{
+		EventHandler: graph.NewEventHandler(100),
+		graph:        g,
+	}
+	c.containerIndexer = newObjectIndexer(g, c.EventHandler, "container", "Namespace", "Pod")
+	c.KubeCache = NewKubeCache(restClient, objType, resources, c)
+	return c
 }
 
-func (p *containerProbe) Stop() {
-	p.graph.RemoveEventListener(p)
+func newContainerProbe(clientset *kubernetes.Clientset, g *graph.Graph) Subprobe {
+	return newContainerCache(clientset.CoreV1().RESTClient(), &v1.Pod{}, "pods", g)
```

Comment:
IIUC, 2 `KubeCache`s are created for pod. That implies 2 HTTP connections/cache/indexers. We should reuse the one use by the "pod" probe

Reply:
inline function

regarding the 2nd commend as I want different handlers (OnAdd/OnUpdate/OnDelete) per each (pod and container) I can't easily just re-used this connection; to implement I would need to create the ability to chain handlers to a single KubeCache which is something a really prefer not to do. If in future we see this pattern (of reuse KubeCache) by additional Subprobes then maybe its' worth the extra complexity and investment.

Reply:
I will try to take a deeper look at how this can be done. But there can be a lot of pods in a Kubernetes cluster so indexing all the pods twice just to create the containers attached to the pods is probably overkill